ci: update Kube and test workflows

wyardley opened the pull request:
#### What this PR does / why we need it:

This just updates and standardizes the Kube versions.

- Update Kubernetes versions and standardize across test matrices
- Update helm-docs to v1.14.2
- Update actions/checkout to v4 in two workflows
- Update helm/chart-testing-action to v2.6.1
- Update setup-helm action to 4.x (4.x is just for node version update)
- Update helm/kind-action to v1.10.0
- Update kubeconform to v0.6.7

See #1199. It had been mentioned there or in another place that 1.16 was still supported - wondering if 1.16 or 1.18 still need to be tested against (I can remove them if wanted).

#### Which issue this PR fixes
*(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close that issue when PR gets merged)*
  - fixes #

#### Special notes for your reviewer:

#### Checklist
[Place an '[x]' (no spaces) in all applicable fields. Please remove unrelated fields.]
- [ ] Chart Version bumped
- [ ] Documentation has been updated with helm-docs (run: `.github/helm-docs.sh`)
- [ ] `CHANGELOG.md` has been updated
- [ ] Variables are documented in the `README.md`
- [ ] For Datadog Operator chart or value changes update the test baselines (run: `make update-test-baselines`)

wyardley commented:
@clamoriniere sorry to bug you directly, but is the codeowners working properly? Doesn't seem to have flagged a person or team on this.

wyardley commented:
@clamoriniere @loutPhilipps @DataDog/container-helm-chart-maintainers - any chance someone could take a look at this one?

clamoriniere reviewed (commented):
The change looks good, I need to run some tests to be sure that the execution works as expected,

wyardley commented:
Thanks. Yeah, I tested one script locally, but obviously wasn’t able to run all the checks so there may be some little tweaks needed.

clamoriniere commented on .github/workflows/ci.yaml:
```diff
@@ -87,10 +87,15 @@ jobs:
+          - v1.26.15
+          - v1.27.16
+          - v1.28.14
+          - v1.29.9
```
I tested this change and it seems that the image 1.29.9 is not available 

see test error: https://github.com/DataDog/helm-charts/actions/runs/11447909017/job/31850338128 

1.29.8 is the last 1.29.x available 

```
➜ crane ls kindest/node | grep 1.29` 
v1.29.0
v1.29.1
v1.29.2
v1.29.4
v1.29.7
v1.29.8
```

clamoriniere commented on .github/workflows/ci.yaml:
```diff
@@ -114,25 +119,30 @@ jobs:
+          - v1.27.16
+          - v1.28.14
+          - v1.29.9
+          - v1.30.5
```
same for 1.35.5

```
➜ crane ls kindest/node | grep 1.30
v1.30.0
v1.30.2
v1.30.3
v1.30.4
```

wyardley replied:
Interesting. The schemas have been present for a bit. 
https://github.com/yannh/kubernetes-json-schema/tree/master/v1.29.9
https://github.com/yannh/kubernetes-json-schema/tree/master/v1.30.5

so maybe they just haven't built those as Docker images yet...

I can just pin all the versions to the same set of versions across both vs. doing new ones for the schemas and older ones for the docker tests?

clamoriniere commented on .github/workflows/ci.yaml:
```diff
@@ -114,25 +119,30 @@ jobs:
+          - v1.26.15
+          - v1.27.16
+          - v1.28.14
+          - v1.29.9
```
I tested this change and it seems that the image 1.29.9 is not available 

see test error: https://github.com/DataDog/helm-charts/actions/runs/11447909017/job/31850338128 

1.29.8 is the last 1.29.x available 

```
➜ crane ls kindest/node | grep 1.29` 
v1.29.0
v1.29.1
v1.29.2
v1.29.4
v1.29.7
v1.29.8
```

clamoriniere commented on .github/workflows/ci.yaml:
```diff
@@ -114,25 +119,30 @@ jobs:
-          - v1.26.13
+          - v1.26.15
+          - v1.27.16
+          - v1.28.14
```
also `v1.28.13` is the last `kindest/node:v1.28.*` version available

clamoriniere reviewed (changes requested):
could you please update the k8s version to use exiting version for the kind/node images

wyardley commented:
@clamoriniere thanks! Rebased, and updated all the tags to get rid of those 3 versions that haven't had Docker images built yet.

Let me know if you want the changed GHA versions to be pinned to hashes vs. versions as well, and if you prefer any further updates to be pushed as separate commits or squashed into one.

wyardley commented:
> could you please update the k8s version to use exiting version for the kind/node images

Yeah, updated. Probably good to just use the same list everywhere?

Within the single file, I guess they could also use a yaml anchor to be a little DRYer, but not sure if that's overkill.

clamoriniere reviewed (approved):
Thanks for the quick update. 
Looks good now 🙇

clamoriniere commented:
/merge

clamoriniere commented:
/merge
